### ISSUE TITLE
FOnlineSessionEpic::FindSessions doesn't correctly cast from FUniqueNetId to FUniqueNetIdEpic

### ISSUE THREAD

**Issue by Garashka** [Ruhrpottpatriot/OnlineSubsystemEpic#26]
# Environment
```none
Project Version: 0.03
EOS SDK Version: 1.7.0
Unreal Engine Version: 4.25.1

Other software/plugins that *might* be the cause of the issue?
```

# Steps to reproduce

1. Run as DebugGame or Standalone
2. Log in successfully using EpicOnlineServicesLoginTask set to Developer
3. Call FOnlineSessionEpic::FindSessions

# Expected behaviour

epicNetId should cast from FUniqueNetId to FUniqueNetIdEpic and retain the epicAccountId from SearchingPlayerId

# Actual behaviour

epicNetId has an epicAccountId of null, even when the epicAccountId value exists on SearchingPlayerId (before it is cast to FUniqueNetIdEpic). This causes the subsystem call to fail and return.

SearchingPlayerId:
![image](https://user-images.githubusercontent.com/8943296/86547809-8ae06380-bf7d-11ea-87ed-99356b33cfed.png)

epicNetId:
![image](https://user-images.githubusercontent.com/8943296/86547791-7ac88400-bf7d-11ea-95fe-f7b0bd87a34d.png)

**Comment by xuelongmu** [comment]
I'm having the exact same issue on UE 4.24.3.

Here you can see the netId (which is passed in as SearchingPlayerId, of type FUniqueNetId):
![](https://cdn.discordapp.com/attachments/710371393321042050/729749888597884938/unknown.png)

And after the cast to FUniqueNetIdEpic, the PUID field is populated but the EAID field is null:
![](https://cdn.discordapp.com/attachments/710371393321042050/729750107091632220/unknown.png)

**Comment by Ruhrpottpatriot** [comment]
I can confirm the bug on my end. The parameter has the  appropriate EAID, but after the cast the EAID is NULL.
Possible reason: Either the copy assignment operator, or the cast doesn't copy the EAID

**Comment by Ruhrpottpatriot** [comment]
So the reason why this wasn't working is the explicit cast operator, that transformed a generic `FUniqueNetId` into an `FUniqueNetIdEpic`:

```
explicit FUniqueNetIdEpic(const FUniqueNetId& OtherId)
    : Type(EPIC_SUBSYSTEM)
    , epicAccountId(nullptr)
{
    // Construct a PUID from the string representation of OtherId
    // Only guarantees grammatical correctness
    FString idString = OtherId.ToString();
    check(!idString.IsEmpty());

    EOS_ProductUserId puid = ProductUserIDFromString(idString);
    check(EOS_ProductUserId_IsValid(puid));
    
    this->productUserId = puid;
}
```

With this implementation the other id is transformed into an `FUniqueNetIdEpic` by converting it to a string and then constructing the new instance from it. However, the `ToString()` method only returns the ProductUserId. This is was a deliberate design choice as there are some situations where one might only want to use the PUID. Additionally the PUID is the centre piece of any unique net id and is always needed, while an EAID is not. [1]

With this problems, the implementation was changed to this:
```
explicit FUniqueNetIdEpic(const FUniqueNetId& OtherId)
	: Type(EPIC_SUBSYSTEM)
	, epicAccountId(nullptr)
{
	if (OtherId.GetType() == EPIC_SUBSYSTEM)
	{
		uint8 const* bytes = OtherId.GetBytes();

		EOS_ProductUserId puid = EOS_ProductUserId_FromString((char const*)bytes);
		check(EOS_ProductUserId_IsValid(puid));
		this->productUserId = puid;

		// If the size of the other id is greater than a PUID's max length
		// we assume that OtherId has a valid EAID
		if (OtherId.GetSize() > EOS_PRODUCTUSERID_MAX_LENGTH)
		{
			EOS_EpicAccountId eaid = EOS_EpicAccountId_FromString((char const*)(bytes + EOS_PRODUCTUSERID_MAX_LENGTH));
			check(EOS_EpicAccountId_IsValid(eaid));
			this->epicAccountId = eaid;
		}
	}
	else
	{
		// Construct a PUID from the string representation of OtherId
		// Only guarantees grammatical correctness
		FString idString = OtherId.ToString();
		check(!idString.IsEmpty());

		EOS_ProductUserId puid = ProductUserIDFromString(idString);
		check(EOS_ProductUserId_IsValid(puid));

		this->productUserId = puid;
	}
}
```

This first checks if the passed `FUniqueNetId` is convertible to the `FUniqueNetIdEpic` type. After that we get the byte representation of the other id. For `FUniqueNetIdEpic` the `GetBytes()` method returns an array of length `EOS_PRODUCTUSERID_MAX_LENGTH + (IsValid(EAID) ? EOS_EPICACCOUNTID_MAX_LENGTH : 0)`. The array has the PUID at position `0` and the EAID at position `EOS_PRODUCTUSERID_MAX_LENGTH`, if the EAID is valid. Therefore we can take the whole byte array and access it at the specified positions and recreate the respective IDs from it.

Only if the type is a different type, the code falls back to using a string representation, as we can assume that no EAID exists and we only want to have a valid PUID.






1: In the future there might be a direct was to convert an `FUniqueNetIdEpic` to a string representation. For now it's possible to get a string representation by converting the id to an EAID via the `ToEpicAccountId` and then use the `EpicAccountIdToString`method.

**Comment by Ruhrpottpatriot** [comment]
Fixed with 93964597886c28078aac9514dc1175ffd74f0973
